### ISSUE TITLE
Two issues of IIS Express v10 on microsoft/windowsservercore:10.0.14393.576

### ISSUE THREAD

**Issue by xied75** [Microsoft/Virtualization-Documentation#524]
I was trying to install IIS Express v10 on the latest (as of 27/12/2016) windowsservercore image:

```
RUN powershell -Command \
    wget https://download.microsoft.com/download/C/E/8/CE8D18F5-D4C0-45B5-B531-ADECD637A1AA/iisexpress_amd64_en-US.msi -OutFile iisexpress_amd64_en-US.msi; \
    Start-Process -FilePath msiexec.exe -ArgumentList /i, "C:\iisexpress_amd64_en-US.msi", /qn, /L, iisexpress_amd64_en-US.install.log -PassThru -Wait;
```

Then looking at the installation log, we got errors like this:

```
Action start 15:05:49: InstallFinalize.
CAQuietExec:  Error: Couldn't add 'IIS Express Application Compatibility Database for x64' to registry.
CAQuietExec:  Uninstallation of IIS Express Application Compatibility Database for x64 complete.
CAQuietExec:  Error 0x80070001: Command line returned an error.
CAQuietExec:  Error 0x80070001: CAQuietExec Failed
CustomAction CA_SETUP_APPCOMPAT returned actual error code 1603 but will be translated to success due to continue marki
g
CAQuietExec:  Error: Couldn't add 'IIS Express Application Compatibility Database for x86' to registry.
CAQuietExec:  Uninstallation of IIS Express Application Compatibility Database for x86 complete.
CAQuietExec:  Error 0x80070001: Command line returned an error.
CAQuietExec:  Error 0x80070001: CAQuietExec Failed
CustomAction CA_SETUP_APPCOMPAT32 returned actual error code 1603 but will be translated to success due to continue mar
ing
```

But on a pretty old TP5 image we don't have this, instead the log look like this:

```
Action start 13:34:59: InstallFinalize.
CAQuietExec:  Installation of IIS Express Application Compatibility Database for x64 complete.
CAQuietExec:  Installation of IIS Express Application Compatibility Database for x86 complete.
CAQuietExec:  Command 'setupDefaultUrls' completed.
```

After looking at the msi with Orca, if I run manually:

```
PS C:\> sdbinst.exe -q .\iisexpressshim.sdb
Error: Couldn't add 'IIS Express Application Compatibility Database for x64' to registry.
Uninstallation of IIS Express Application Compatibility Database for x64 complete.
```

So this is issue one, that we got incomplete install of IIS Express here.

Now issue two is even more of a trouble, for an asp.net site I did which works perfectly before on TP5, now if I do ```curl http://localhost -UseBasicParsing``` I would get

```
Server Error in '/' Application.
 Specified argument was out of the range of valid values.Parameter name: site
```

This seems can be solved by
```
PS C:\> Install-WindowsFeature Web-WHC
```
as mentioned here: https://social.msdn.microsoft.com/Forums/windows/en-US/cb6a04e1-19d6-4a75-be25-aca16b89f166/server-error-application-httpexception-0x80004005?forum=winformssetup

But nonetheless it does work before on TP5, without needing this extra bit, which kind of cancelled the point of having an Express. So it must be something in the base image for these two issues.

**Comment by friism** [comment]
@xied75 the `Web-WHC` is probably a change in default from TP5, you should add it as part of the Dockerfile if you need it.

I can repro the msi problem. It doesn't look like there's an MSI parameter to turn of that feature.

cc @PatrickLang 

@xied75 out of curiosity, why are you wanting to run IIS express vs. full IIS (for which there are Microsoft-maintained base images).

**Comment by xied75** [comment]
@friism I went back to TP5 and checked, there is no WHC enabled there. So it's something else. In any way IIS Express should work by itself without rely on some magic components, nothing mentioned in its documentation for this.

Regarding the msi problem, it does not affect the website running; the compatibility database is just a bunch of HKLM keys, but it might worth finding out why it doesn't work on RTM base image, as it could affect other softwares.

For the last question, I'm using IIS Express for this:
https://github.com/coderobin/windows-container/tree/master/windowsservercore/nuget.server
Given this is more a "workgroup" solution, as NuGet team also has a full blown version which is better suited for full IIS, I thought it's reasonable to use the "smaller" thing for it.

Besides, IIS Express is an executable, not a service, making it much better fit with Docker idiom, look the difference of my ENTRYPOINT with the official IIS one, the "ping" thing is not ideal. If the service die, the container should stop.

**Comment by PatrickLang** [comment]
Ping @shirhatti - do you know someone that could help with this one? I'm not familiar with IIS Express. We should get a code sample published in this doc repo

**Comment by mylesbkeating1993** [comment]
@PatrickLang @xied75 

Is there anything I can help with?  Trying to close dead issues

**Comment by xied75** [comment]
I'm still waiting answers from you guys.

**Comment by mylesbkeating1993** [comment]
OK I circled back with the team and got some answers.  Sorry this has been sitting so long.  That's our bad.

There was a bug.  You helped us find it, thank you! 

We fixed the bug and shipped the patch with an OS update a few months ago.  If you still are having issues, definitely let me know!

If you are curious, the bug was around the way containers handle registry keys.


I was looking at your [NuGet project](https://github.com/coderobin/windows-container/commit/afea5f1252dd86d73a8107de5ed50b7e4aaf93b8) and I want to make sure I'm helping you achieve your goals with Windows Containers.  What are you trying to do exactly?

**Comment by xied75** [comment]
@mylesbkeating1993 Thanks for all these.

Regarding the fix, do you mean the fix is on the container host (i.e. Windows Server 2016 in my case) or the windowsservercore base image or both?

I was using 10.0.14393.576 when I raised this issue. Could you kindly point out from which version the fix is in? https://hub.docker.com/r/microsoft/windowsservercore/tags/

About why I'm using IIS Express with this simple NuGet project:

1, This is one component in my CI build pipeline. My Jenkins build will be triggered by GitHub push and build my portable libraries and push to my Nuget source.
2, when build containers the best practice is to keep it as small as possible, in my case I just need a nuget container that can hold enough packages for my UWP app development use, therefore using the "proper" https://github.com/NuGet/NuGetGallery is overkill, thus using full IIS is also overkill.

**Comment by mylesbkeating1993** [comment]
Ok update.  
Sorry for the delay - DockerCon and prepping for Build have taken over our lives haha

I jumped the gun, it's a little more complicated.

- The fix is in the OS, not the image
- It has to do with registry key virtualization.  Happy to explain further if you are curious
- Original plan was to create and ship a fix with Server 2016, but the work was pushed back on the grounds that using full IIS is a workaround
- Since this is a Server OS fix, this means the next opportunity to ship it is actually in the fall

So in short, if you want to use IIS express you'll need to wait for the Server update in the fall.  In the meantime, full IIS is your best bet.

I understand this isn't your ideal scenario to make your container as small as possible.  Unfortunately we have more things to work on than we have time to complete, so we need to prioritize.

Please let me know if I can help with anything else!

Again, thanks a lot for posting.  You helped us find a bug and a fix is on the way :)

**Comment by xied75** [comment]
@mylesbkeating1993 Thanks a lot for your good work and good words. I fully understand how much backlog you guys have for Windows container. Keep going.

Wish you a great Build, (are you giving any talks? Which session?) 

Feel free to close this issue. :)

p.s. This is the App I built with my own CI pipeline fully running in Windows Containers on Azure:
https://www.microsoft.com/store/apps/9ntphrk6qlfg

**Comment by mylesbkeating1993** [comment]
@xied75 Thank you for using our stuff and building cool things with it!

Code Reader looks awesome :)

Taylor is indeed presenting tomorrow with Steve Lasker: https://channel9.msdn.com/Events/Build/2017/B8013

**Comment by edalbez** [comment]
hi,
Why did you thy this ?
PS C:\> sdbinst.exe -q .\iisexpressshim.sdb

**Comment by xied75** [comment]
@edalbez This is merely a step to reproduce the error message, to show where the error occurs during a multi-step installation of a MSI file. Within the iisexpress_amd64_en-US.msi file, there is a step doing the same action like this command line. Orca is a tool to open up MSI to look inside. If you got further questions, feel free to email me directly.
